src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs

```diff
@@ -0,0 +1,196 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// Helper type for building dynamically-sized arrays while minimizing allocations and copying.
+    /// </summary>
+    internal struct LargeArrayBuilder<T>
+    {
+        private const int StartingCapacity = 4;
+        private const int ResizeLimit = 32;
+
+        private T[] _first;                 // The first buffer we store items in. Resized until ResizeLimit.
+        private ArrayBuilder<T[]> _buffers; // After ResizeLimit * 2, we store previous buffers we've filled out here.
+        private T[] _current;               // Current buffer we're reading into. Is _first <= ResizeLimit.
+        private int _index;                 // Index into the current buffer.
+        private int _count;                 // Count of all of the items in this builder.
+
+        /// <summary>
+        /// Constructs a new builder.
+        /// </summary>
+        /// <param name="initialize">Pass <c>true</c>.</param>
+        public LargeArrayBuilder(bool initialize) : this()
```

Comment:
> along with the fact that readability is helped by seeing common patterns and harmed by seeing jarring patterns, and to me, this use of ": this()" in a struct is jarring.

OK, I can remove the explicit `: this()` call then. Sorry.

> A compiler can choose to optimize certain patterns, so the compiler could choose to generate the exact same code for:

Yes. I was trying to say that with the first version the JIT would (probably) have to inspect more bytecode to see that `0` was being assigned to `_field1`, and `null` was being assigned to `_field2`, before making that optimization (which would presumably come at a cost since all of this is done at runtime).